pkg/apis/kubebind/v1alpha1/permissionclaims_types.go

```diff
@@ -0,0 +1,273 @@
+	// both sides.
+	//
+	// autoAdopt is mutually exclusive with autoDonate.
+	//
+	// +optional
+	AutoAdopt bool `json:"autoAdopt,omitempty"`
+
+	// autoDonate set to true means that a newly created object synced from the
+	// provider to the consumer cluster is immediately donated to the consumer
+	// by marking it as owned by the consumer on both sides.
+	//
+	// autoDonate is mutually exclusive with autoAdopt.
+	//
+	// +optional
+	AutoDonate bool `json:"autoDonate,omitempty"`
+
+	// onConflict determines how conflicts between objects on the consumer
+	// and provider clusters will be resolved.
+	//
+	// +optional
+	// +kubebuilder:default:={}
+	OnConflict *PermissionClaimOnConflictOptions `json:"onConflict,omitempty"`
+
+	// update lists which updates to objects on the consumer cluster are claimed.
+	// By default, the whole object is synced, but metadata is not.
+	//
+	// Note that update permissions do not imply create permissions.
+	//
+	// +optional
+	Update *PermissionClaimUpdateOptions `json:"update,omitempty"`
+}
+
+type PermissionClaimReadOptions struct {
+	// labels is a list of label key wildcard patterns that are synchronized
+	// from the consumer to the provider on consumer-owned objects.
+	//
+	// +optional
+	Labels []Matcher `json:"labels,omitempty"`
+
+	// LabelsOnProviderOwnedObjects is a list of claimed label key wildcard
+	// patterns that are synchronized from the consumer cluster to the provider
+	// on objects owned by the provider.
+	//
+	// +optional
+	LabelsOnProviderOwnedObjects []Matcher `json:"labelsOnProviderOwnedObjects,omitempty"`
+
+	// annotations is a list of claimed annotation key wildcard patterns
+	// that are synchronized from the consumer cluster to the provider on
+	// objects that are owned by the consumer.
+	//
+	// +optional
+	Annotations []Matcher `json:"annotations,omitempty"`
+
+	// overrideAnnotations is a list of claimed annotation key wildcard
+	// patterns that are synchronized from the consumer cluster to the provider
+	// on provider-owned objects.
+	//
+	// +optional
+	AnnotationsOnProviderOwnedObjects []Matcher `json:"annotationsOnProviderOwnedObjects,omitempty"`
+}
+
+type Matcher struct {
+	// pattern is a wildcard pattern that is matched against the key. This means
+	// it is either a literal string or starts or ends in `*` but is not '*'
+	// itself.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength:=1
+	Pattern string `json:"pattern,omitempty"`
+}
+
+type PermissionClaimOnConflictOptions struct {
+	// recreateWhenConsumerSideDeleted set to true (the default) means the provider will recreate the object
+	// in case the object is missing on the consumer cluster, but has been synchronized before.
+	//
+	// If set to false, deleted provider-owned objects get deleted on the provider cluster as well.
+	//
+	// +kubebuilder:default:=true
+	RecreateWhenConsumerSideDeleted bool `json:"recreateWhenConsumerSideDeleted"`
```

Comment:
These looks like they will be mutual exlusive too, maybe something similar to:
```
// Similarly to above, these are constants to support HTTP PATCH utilized by
// both the client and server that didn't make sense for a whole package to be
// dedicated to.
type PatchType string

const (
	JSONPatchType           PatchType = "application/json-patch+json"
	MergePatchType          PatchType = "application/merge-patch+json"
	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
	ApplyPatchType          PatchType = "application/apply-patch+yaml"
)
```
As I bet these will be more here, like `ForceRecreate`, `DeleteAndRecreate`, etc

Reply:
you write above to move this into update or create. Neither is really a fit.

Reply:
Maybe `create.onDelete: Recretate|DeleteOnProviderSide`.

Reply:
I’m not sure if it was brought up in the previous discussions, but what about deletion scenarios? Currently an object removed from the consumer cluster is triggering deletion on the provider side. If you remove the object on the provider side, it will be re-created from the consumer copy. Meaning that in the `Recreate` state there is no way to remove the object from the cluster without stopping the konnector instance. Shouldn’t it be more general? `Recreate|FollowDeletion` allowing for both sides being equal in having right for resource eviction?

Reply:
That's where the `onConflict` part of the field name comes from. The action is performed when the lifecycle of the object conflicts with the state on the cluster.

> Meaning that in the Recreate state there is no way to remove the object from the cluster without stopping the konnector instance
For provider owned objects, that is true (short of undoing the binding). Claimed resources are objects that support objects of a bound api. For example, creating a `MangoDB` object should spawn a secret containing credentials to use the MangoDB instance. It is expected that the supporting objects are removed upon deletion of the 'main' resource.

The behavior in case of deletion looks like the following:

## Recreate: false

**provider** owned object deleted: not recreated on the consumer side. Since the provide still manages the object, the provider will eventually delete it on the provider cluster.

**consumer** owned object: the object will be deleted on the provider cluster, since the consumer object controls the lifecycle

## Recreate: true

**provider** owned object: the object is created again on the consumer cluster, once the provider deletes it on the provider cluster, it is deleted on the consumer cluster

**consumer** owned object: the object will be deleted on the provider cluster, since the consumer object controls the lifecycle

Reply:
maybe onConflict.userDeletesProviderOwnedObject = Recreate | Orphan?

Reply:
So an object created on the provider side and synchronized on the consumer side will behave the same way as the object created on the consumer side and synchronized on the provider side is idempotent when it comes to deletion? Basically the `FollowDeletion` or `Follow` scenario, is supported by the api?

Reply:
An object that is created by the provider and synchronized to the consumer could be donated using `ownerTransfer=Donate`. In that case, if the consumer deletes it, it will be deleted on the provider cluster too. Since donation means that the consumer will own the object, the provider may have to claim update fields to continue writing to the object after creation in the consumer cluster